### PR TITLE
[WOR-1186] Async workspace deletion API wireup

### PR DESCRIPTION
**Pull request by aherbst-broad**
Ticket: https://broadworkbench.atlassian.net/browse/WOR-1186

## Context
This is the first step in the effort to build out asynchronous azure workspace deletion. This builds out the needed API endpoint and service modifications up to the point of depositing a monitor record in the workspace manager monitoring table. [WOR-1193](https://broadworkbench.atlassian.net/browse/WOR-1193) will pick up where this leaves off an implement the backend job to delete a workspace.

## This PR
* Adds a new versioned workspace endpoint:
   ```DELETE /api/workspaces/v2/<namespace>/<name>```
* Adds a new method to the `MultiCloudWorkspaceService`
   * When presented with an "MC" workspace, it creates the needed record on the monitor table and gives back the job ID
   * When presented with a "rawls" workspace, deletes the workspace synchronously and gives back the bucketName.
 * Introduces a `WorkspaceDeletionResult` type. For now this will contain the jobId for the async path. 
    * I've added a `gcpDeletionContext` field that will contain the `bucketName` for some semblance of compatibility with clients. This is not fully backwards compatible with the existing API as the v1 API actually [renders](https://github.com/broadinstitute/rawls/blob/develop/core/src/main/scala/org/broadinstitute/dsde/rawls/webservice/WorkspaceApiService.scala#L291) a user facing message and returns that to the UI. My hunch is to give back just the needed metadata and let the rendering of a user facing string be a UI concern.

---

**PR checklist**

- [ ] Include the JIRA issue number in the PR description and title
- [ ] Make sure Swagger is updated if API changes
  - [ ] **...and Orchestration's Swagger too!**
- [ ] If you changed anything in `model/`, then you should [publish a new official `rawls-model`](https://github.com/broadinstitute/rawls/blob/develop/README.md#publish-rawls-model) and update `rawls-model` in [Orchestration's dependencies](https://github.com/broadinstitute/firecloud-orchestration/blob/develop/project/Dependencies.scala).
- [ ] Get two thumbsworth of PR review
- [ ] Verify all tests go green, including CI tests
- [ ] **Squash commits and merge** to develop (branches are automatically deleted after merging)
- [ ] Inform other teams of any substantial changes via Slack and/or email

[WOR-1193]: https://broadworkbench.atlassian.net/browse/WOR-1193?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

**Comment by aherbst-broad** [comment]
jenkins retest

**Comment by aherbst-broad** [comment]
jenkins retest

**Review by blakery — APPROVED** [review]

File: model/src/main/scala/org/broadinstitute/dsde/rawls/model/WorkspaceModel.scala [hunk 1]
```diff
@@ -148,6 +148,19 @@ object WorkspaceVersions {
     )
 }
 
+case class GcpWorkspaceDeletionContext(bucketName: Option[String])
```
Comment (marctalbott) [inline model/src/main/scala/org/broadinstitute/dsde/rawls/model/WorkspaceModel.scala]: `bucketName` can be unwrapped from the `Option`. All GCP workspaces will have buckets, and Azure workspace deletion results won't have a `GcpWorkspaceDeletionContext`.

File: core/src/main/scala/org/broadinstitute/dsde/rawls/workspace/MultiCloudWorkspaceService.scala [hunk 2]
```diff
@@ -109,13 +112,62 @@ class MultiCloudWorkspaceService(override val ctx: RawlsRequestContext,
...
+        s"V2 Starting deletion of workspace [workspaceId=${workspace.workspaceId}, name=${workspaceName.name}, billingProject=${workspace.namespace}, user=${ctx.userInfo.userSubjectId.value}]"
+      )
+
+      _ = workspace.state match {
+        case WorkspaceState.Ready => true
+        case WorkspaceState.Deleting =>
+          throw new RawlsExceptionWithErrorReport(
+            ErrorReport(StatusCodes.Conflict, "Workspace is already being deleted.")
+          )
+        case _ =>
```
Comment (marctalbott) [inline core/src/main/scala/org/broadinstitute/dsde/rawls/workspace/MultiCloudWorkspaceService.scala:137]: Do we need to allow workspaces in `CreateFailed`, `DeleteFailed`, and `UpdateFailed` to be deleted so users can clean up their failed workspaces?

**Review by marctalbott — APPROVED** [review]

File: core/src/main/scala/org/broadinstitute/dsde/rawls/workspace/MultiCloudWorkspaceService.scala (hunk 2 above)
Comment (blakery) [reply core/src/main/scala/org/broadinstitute/dsde/rawls/workspace/MultiCloudWorkspaceService.scala:137]: That's a good point.

I'm trying to think through the implications of that. We might need to make some steps more resilient to things like incomplete records or resources, for that to work.
On the other hand, if deleting doesn't work, it just goes from one failed state to another? 

Maybe we should allow this after the 'Nuke it from orbit' plan is in place?

Comment (cahrens) [reply core/src/main/scala/org/broadinstitute/dsde/rawls/workspace/MultiCloudWorkspaceService.scala:137]: I think we should allow deletion for any state besides Deleting.

Comment (aherbst-broad) [reply core/src/main/scala/org/broadinstitute/dsde/rawls/workspace/MultiCloudWorkspaceService.scala:137]: I will update. We don't want to admit transitions from CREATING and UPDATING (per the [design spec](https://broadworkbench.atlassian.net/wiki/spaces/WOR/pages/2819653639/Workspace+Async+Deletion)) so I will fail those out.

**Comment by aherbst-broad** [comment]
jenkins retest

**Comment by aherbst-broad** [comment]
jenkins retest

**Comment by aherbst-broad** [comment]
jenkins retest

**Pull request merged**